### ISSUE TITLE
Pyright does not detect mutated attribute types

### ISSUE THREAD

**Issue by Archmonger** [microsoft/pyright#10013]
**Describe the bug**
When doing standard Python programing practices to ensure an attribute is a certain type within the `__init__` method, pyright does not detect the user's type validation/mutation and decides the original type takes precedence at all times.

**Code or Screenshots**
For example...

```python
from dataclasses import dataclass

@dataclass
class DataClassExample:
    name: str

data_class = DataClassExample("John")

class Example:
    def __init__(self, name):
        self.name: str | None = getattr(data_class, "name", None)

        if self.name is None:
            self.name = "Hello"

    def __call__(self):
        return self.name + " World"
        #      ^^^^^^^^^
        # error: Operator "+" not supported for "None" (reportOptionalOperand)
```

In the scenario above, pyright should be able to infer that `self.name` will always be a string. However, it instead provides an `reportOptionalOperand` error.

The same error occurs if substituting `self.name is None` with an `isinstance` check.

```python
        ...
        if not isinstance(name, str):
            self.name = "Hello"
        ...
```

Rationally, the logical step to fix this would be to redefine the type hint during the type check, but that results in a whole different error.

```python
class Example:
    def __init__(self, name):
        self.name: str | None = getattr(data_class, "name", None)
        # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        # error: Declaration "name" is obscured by a declaration of the same name (reportRedeclaration)
        # error: Cannot assign to attribute "name" for class "Example*".  Type "Any | None" is not assignable to type "str".  "None" is not assignable to "str" (reportAttributeAccessIssue)

        if not isinstance(name, str):
            self.name: str = "Hello"

    def __call__(self):
        return self.name + " World"
```

**VS Code extension or command-line**
pyright v1.1.395

**Comment by erictraut** [comment]
Pyright is working correctly here, so this isn't a bug.

You have declared the type of the `name` attribute to be `str | None`. That means it's legal to assign it a value of `None` at any point.

```python
e = Example("")
e.name = None
e() # Crash
```

If you want `name` to contain only a `str`, you should change the type declaration to `str` and make sure you never assign `None` to it.

```python
    def __init__(self, name):
        self.name: str = getattr(data_class, "name", "Hello")
```
